### ISSUE TITLE
Installation fails on Ubuntu Ubuntu 24.04.1 LTS

### ISSUE THREAD

**Issue by GertVersteeg1956** [jasonacox/tinytuya#555]
When I try to install the library on Ubuntu 24.04.1 LTS it Fails.

> python -m pip install tinytuya
> error: externally-managed-environment
> 
> × This environment is externally managed
> ╰─> To install Python packages system-wide, try apt install
>     python3-xyz, where xyz is the package you are trying to
>     install.
>     
>     If you wish to install a non-Debian-packaged Python package,
>     create a virtual environment using python3 -m venv path/to/venv.
>     Then use path/to/venv/bin/python and path/to/venv/bin/pip. Make
>     sure you have python3-full installed.
>     
>     For more information visit http://rptl.io/venv
> 
> note: If you believe this is a mistake, please contact your Python installation or OS distribution provider. You can override this, at the risk of breaking your Python installation or OS, by passing --break-system-packages.
> hint: See PEP 668 for the detailed specification.

The first option does not work either

> sudo apt install python3-tinytuya

> [sudo] password for gert: 
> Pakketlijsten worden ingelezen... Klaar
> Boom van vereisten wordt opgebouwd... Klaar
> De statusinformatie wordt gelezen... Klaar 
> E: Kan pakket python3-tinytuya niet vinden

Can you please advise how to proceed

**Comment by jasonacox** [comment]
Ubuntu 24.04.1 LTS prevents you from installing python libraries. This is the new default behavior because they want you to create virtual environments instead of changing the default configuration.  TinyTuya is not a Debian packaged python so you will need to proceed with the second option with something like this:

```bash
python3 -m venv ~/.venv
source ~/.venv/bin/activate
pip install tinytuya
```

This builds a copy of python in ~/.venv with all the packages under your control without messing up the system packages. 
You will need to add that source to your .bashrc so it activates every time, or make explicit calls to ~/.venv/bin/python3 to run your scripts.

TODO: 
* I should explore how we can get a "debian packaged" version of tinytuya to make the first option work. If anyone has experience with that, please chime in. :) 
* I should update documentation since many of the Linux variants are moving to this method

**Comment by GertVersteeg1956** [comment]
Hi Jason
Thanks for your quick response. I'll proceed with the second option and keep you informed.
Kind regards
Gert

Jason Cox ***@***.***> schreef op 9 november 2024 07:10:09 CET:

**Comment by GertVersteeg1956** [comment]
I managed to install it in a virtual environment. Next step is to play with the library and integrate it in a copy of xenomes' local tuya plugin.

**Comment by GertVersteeg1956** [comment]
After installing it, I downloaded Xenomes' package https://github.com/Xenomes/Domoticz-TinyTUYA-Local-Plugin and copied the tinytuya directory, library to the plugin directory.
I copied the devices.json file from previous experiments and adjusted the IPV4 adresses in this file.
When I start the plugin it sees my 3 lsc power plugs (action) and my dehumidifier (qlima). Xenomes helped me before with this ;-)
It creates various devices, but does not update the values; I can switch the powerplugs and dehumidifier on and off from domoticz.
I'll use this plugin to study python domoticz plugin mechanism. 
I can keep you posted when you are interested; I can also create a domoticz forum or tweakers topic. Don´t know yet what the best way is to inform others about my findings.
